MCOL-4590 UNION Performance Improvement with the focus on the normalize functions.

JigaoLuo opened the pull request:
The Jira issue number for this PR is: [MCOL-4590](https://jira.mariadb.org/browse/MCOL-4590)

*NOTE: This project is for the Google Summer of Code 2022.*

# Task

- [x] Simple Approach: use separate functions instead of the huge switch
- [x] Code Review and correction in simple approach

tntnatbry commented on dbcon/joblist/tupleunion.h:
```diff
@@ -122,8 +122,8 @@ class TupleUnion : public JobStep, public TupleDeliveryStep
-  void addToOutput(rowgroup::Row* r, rowgroup::RowGroup* rg, bool keepit, rowgroup::RGData& data);
-  void normalize(const rowgroup::Row& in, rowgroup::Row* out);
+  void addToOutput(rowgroup::Row* r, rowgroup::RowGroup* rg, bool keepit, rowgroup::RGData& data, uint32_t& tmpOutputRowCount);
+  void normalize(const rowgroup::Row& in, rowgroup::Row* out, std::vector<std::function<void(const rowgroup::Row& in, rowgroup::Row* out, uint32_t col)>>& infer_normalize_functions);
```
Please change `infer_normalize_functions` to camelCase: `inferNormalizeFunctions`.

tntnatbry commented on dbcon/joblist/tupleunion.cpp:
```diff
@@ -218,15 +1508,17 @@ void TupleUnion::readInput(uint32_t which)
           memUsageBefore = allocator.getMemUsage();
 
+          uint32_t tmpOutputRowCount = l_outputRG.getRowCount();
           for (uint32_t i = 0; i < l_tmpRG.getRowCount(); i++, tmpRow.nextRow())
```
Can you replace `l_tmpRG.getRowCount()` call in the loop with a constant:
```
const uint32_t tmpRGRowCount = l_tmpRG.getRowCount();
```

tntnatbry commented on dbcon/joblist/tupleunion.cpp:
```diff
@@ -236,12 +1528,15 @@ void TupleUnion::readInput(uint32_t which)
             }
           }
 
+          fRowsReturned += (tmpOutputRowCount - l_outputRG.getRowCount());
```
I don't think this calculation of `fRowsReturned` is correct. What if `tmpOutputRowCount < l_outputRG.getRowCount()`?

Instead, `fRowsReturned` should be moved inside the `if` statement above:

```
if (inserted.second)
{
  copyRow(tmpRow, &outRow);
  const_cast<RowPosition&>(*(inserted.first)) =
    RowPosition(rowMemory.size() - 1, tmpOutputRowCount);
  memDiff += outRow.getRealSize();
  addToOutput(&outRow, &l_outputRG, true, outRGData, tmpOutputRowCount);
  fRowsReturned++;
}
```

tntnatbry commented on dbcon/joblist/tupleunion.cpp:
```diff
@@ -56,6 +56,1296 @@ inline double exp10(double x)
+      {
+        result.emplace_back();  /// If Null, then add a placeholder in the vector.
+        continue;
+      }
```
The above `in.isNullValue(i)` check is incorrect. This check cannot be done at the inferring phase as we don't look at the actual column values during inferring but only the column type attributes. This should be removed.

tntnatbry commented on dbcon/joblist/tupleunion.cpp:
```diff
@@ -56,6 +56,1296 @@ inline double exp10(double x)
+  
+  void normalizeIntToUintNoScale(const Row& in, Row* out, uint32_t i) 
+  {
+    out->setIntField(in.getIntField(i), i); 
```
This should be `out->setUintField(in.getIntField(i), i);`

tntnatbry commented on dbcon/joblist/tupleunion.cpp:
```diff
@@ -56,6 +56,1296 @@ inline double exp10(double x)
+              if (out->getColumnWidth(i) == datatypes::MAXDECIMALWIDTH)
+                result.emplace_back(normalizeIntToXDecimalWithScaleInt128);
+              else
+                result.emplace_back(normalizeIntToXDecimalWithScaleInt64);
```
The above two functions should be renamed to `normalizeIntToXDecimalInt128` and `normalizeIntToXDecimalInt64` as there is no usage of scale here.

tntnatbry commented on dbcon/joblist/tupleunion.cpp:
```diff
@@ -56,6 +56,1296 @@ inline double exp10(double x)
+    const int diff = out->getScale(i) - in.getScale(i);
+    idbassert(diff >= 0);
+    uint64_t val = datatypes::applySignedScale<uint64_t>(in.getUintField(i), diff);
+    out->setUintField(val, i);
```
I think this should be `out->setIntField(val, i);`. This looks like an error in the original code. Function name should also be changed to `normalizeUntToIntWithScaleInt64`.

tntnatbry commented on dbcon/joblist/tupleunion.cpp:
```diff
@@ -56,6 +56,1296 @@ inline double exp10(double x)
+              if (out->getColumnWidth(i) == datatypes::MAXDECIMALWIDTH)
+                result.emplace_back(normalizeUintToXDecimalWithScaleInt128);
+              else
+                result.emplace_back(normalizeUintToXDecimalWithScalUint64);
```
The above two functions should be renamed to `normalizeUintToXDecimalInt128` and `normalizeUintToXDecimalInt64` respectively.

tntnatbry commented on dbcon/joblist/tupleunion.cpp:
```diff
@@ -56,6 +56,1296 @@ inline double exp10(double x)
+    const int diff = out->getScale(i) - in.getScale(i);
+    idbassert(diff >= 0);
+    uint64_t val = datatypes::applySignedScale<uint64_t>(in.getUintField(i), diff);
+    out->setUintField(val, i);
```
This should be `out->setIntField(val, i);`.

tntnatbry commented on dbcon/joblist/tupleunion.cpp:
```diff
@@ -56,6 +56,1296 @@ inline double exp10(double x)
+                result.emplace_back(normalizeDateToTimestampInvalid);
+              else 
+                result.emplace_back(std::bind(normalizeDateToTimestampValid, std::placeholders::_1, std::placeholders::_2, std::placeholders::_3, seconds, m_time.second_part));
+              break;
```
This case is incorrect. There can be no reference to `in.getUintField(i)` during inferring. You can merge the above two functions into one `normalizeDateToTimestamp` and have the if-else condition inside that function.

tntnatbry commented on dbcon/joblist/tupleunion.cpp:
```diff
@@ -56,6 +56,1296 @@ inline double exp10(double x)
+              else 
+                result.emplace_back(std::bind(normalizeDatetimeToTimestampValid, std::placeholders::_1, std::placeholders::_2, std::placeholders::_3, seconds, m_time.second_part));
+              break;
+            }
```
Same as above. This case is incorrect. There can be no reference to `in.getUintField(i)` during inferring. You can merge the above two functions into one `normalizeDatetimeToTimestamp` and have the if-else condition inside that function.

tntnatbry commented on dbcon/joblist/tupleunion.cpp:
```diff
@@ -56,6 +56,1296 @@ inline double exp10(double x)
+                {
+                  if (out->getColumnWidth(i) == datatypes::MAXDECIMALWIDTH)
+                  {
+                    if (in.getColTypes()[i] == CalpontSystemCatalog::FLOAT)
```
This should be `if (in.getColTypes()[i] == CalpontSystemCatalog::FLOAT || in.getColTypes()[i] == CalpontSystemCatalog::UFLOAT)`

tntnatbry commented on dbcon/joblist/tupleunion.cpp:
```diff
@@ -56,6 +56,1296 @@ inline double exp10(double x)
+                  }
+                  else
+                  {
+                    if (in.getColTypes()[i] == CalpontSystemCatalog::FLOAT)
```
This should be `if (in.getColTypes()[i] == CalpontSystemCatalog::FLOAT || in.getColTypes()[i] == CalpontSystemCatalog::UFLOAT)`

tntnatbry commented on dbcon/joblist/tupleunion.cpp:
```diff
@@ -56,6 +56,1296 @@ inline double exp10(double x)
+                } 
+                else
+                {
+                  if (in.getColTypes()[i] == CalpontSystemCatalog::FLOAT)
```
This should be `if (in.getColTypes()[i] == CalpontSystemCatalog::FLOAT || in.getColTypes()[i] == CalpontSystemCatalog::UFLOAT)`

tntnatbry commented on dbcon/joblist/tupleunion.cpp:
```diff
@@ -56,6 +56,1296 @@ inline double exp10(double x)
+    const int diff = out->getScale(i) - scale;
+    ival = datatypes::applySignedScale<uint64_t>(ival, diff);
+    out->setIntField(ival, i);
+  }
```
The following code is duplicated in the above four functions:

```c++
/* have to pick a scale to use for the double. using 5... */
    uint32_t scale = 5;
    uint64_t ival = (uint64_t)(double)(val * datatypes::scaleDivisor<double>(scale));
    const int diff = out->getScale(i) - scale;
    ival = datatypes::applySignedScale<uint64_t>(ival, diff);
```
Can you remove this duplication by creating a common inline function?

tntnatbry commented on dbcon/joblist/tupleunion.cpp:
```diff
@@ -56,6 +56,1296 @@ inline double exp10(double x)
+            case CalpontSystemCatalog::UINT:
+            case CalpontSystemCatalog::UBIGINT: 
+            {
+              if (in.getColTypes()[i] == CalpontSystemCatalog::FLOAT)
```
This should be `if (in.getColTypes()[i] == CalpontSystemCatalog::FLOAT || in.getColTypes()[i] == CalpontSystemCatalog::UFLOAT)`

tntnatbry commented on dbcon/joblist/tupleunion.cpp:
```diff
@@ -56,6 +56,1296 @@ inline double exp10(double x)
+            case CalpontSystemCatalog::FLOAT:
+            case CalpontSystemCatalog::UFLOAT: 
+            {
+              if (in.getColTypes()[i] == CalpontSystemCatalog::FLOAT)
```
This should be `if (in.getColTypes()[i] == CalpontSystemCatalog::FLOAT || in.getColTypes()[i] == CalpontSystemCatalog::UFLOAT)`

tntnatbry commented on dbcon/joblist/tupleunion.cpp:
```diff
@@ -56,6 +56,1296 @@ inline double exp10(double x)
+            case CalpontSystemCatalog::DOUBLE:
+            case CalpontSystemCatalog::UDOUBLE: 
+            {
+              if (in.getColTypes()[i] == CalpontSystemCatalog::FLOAT)
```
This should be `if (in.getColTypes()[i] == CalpontSystemCatalog::FLOAT || in.getColTypes()[i] == CalpontSystemCatalog::UFLOAT)`

tntnatbry commented on dbcon/joblist/tupleunion.cpp:
```diff
@@ -56,6 +56,1296 @@ inline double exp10(double x)
+            
+            case CalpontSystemCatalog::LONGDOUBLE: 
+            {
+              if (in.getColTypes()[i] == CalpontSystemCatalog::FLOAT)
```
This should be `if (in.getColTypes()[i] == CalpontSystemCatalog::FLOAT || in.getColTypes()[i] == CalpontSystemCatalog::UFLOAT)`

tntnatbry commented on dbcon/joblist/tupleunion.cpp:
```diff
@@ -56,6 +56,1296 @@ inline double exp10(double x)
+            case CalpontSystemCatalog::TEXT:
+            case CalpontSystemCatalog::VARCHAR: 
+            {
+              if (in.getColTypes()[i] == CalpontSystemCatalog::FLOAT)
```
This should be `if (in.getColTypes()[i] == CalpontSystemCatalog::FLOAT || in.getColTypes()[i] == CalpontSystemCatalog::UFLOAT)`

tntnatbry commented on dbcon/joblist/tupleunion.cpp:
```diff
@@ -56,6 +56,1296 @@ inline double exp10(double x)
+              // TODO: isn't overflow possible below?
+              if (out->getColumnWidth(i) == datatypes::MAXDECIMALWIDTH)
+              {
+                if (in.getColTypes()[i] == CalpontSystemCatalog::FLOAT)
```
This should be `if (in.getColTypes()[i] == CalpontSystemCatalog::FLOAT || in.getColTypes()[i] == CalpontSystemCatalog::UFLOAT)`

tntnatbry commented on dbcon/joblist/tupleunion.cpp:
```diff
@@ -56,6 +56,1296 @@ inline double exp10(double x)
+              }
+              else              
+              {
+                if (in.getColTypes()[i] == CalpontSystemCatalog::FLOAT)
```
This should be `if (in.getColTypes()[i] == CalpontSystemCatalog::FLOAT || in.getColTypes()[i] == CalpontSystemCatalog::UFLOAT)`

tntnatbry commented on dbcon/joblist/tupleunion.cpp:
```diff
@@ -56,6 +56,1296 @@ inline double exp10(double x)
+    const int diff = out->getScale(i) - scale;
+    ival = datatypes::applySignedScale<uint64_t>(ival, diff);
+    out->setIntField(ival, i);
+  }
```
The following code is duplicated in the above four functions:
```c++
    /* have to pick a scale to use for the double. using 5... */
    uint32_t scale = 5;
    uint64_t ival = (uint64_t)(double)(val * datatypes::scaleDivisor<double>(scale));
    const int diff = out->getScale(i) - scale;
    ival = datatypes::applySignedScale<uint64_t>(ival, diff);
```

Can you please remove this duplication by creating an inline function? This is the same code that is also duplicated in four more functions above. See my comment at `normalizeXDoubleToIntWithScaleInt64`.

tntnatbry commented on dbcon/joblist/tupleunion.cpp:
```diff
@@ -56,6 +56,1296 @@ inline double exp10(double x)
+              if (out->getColumnWidth(i) == datatypes::MAXDECIMALWIDTH)
+                result.emplace_back(normalizeLongDoubleToXDecimalWithScaleInt128);
+              else
+                result.emplace_back(normalizeLongDoubleToXDecimalWithScaleInt);
```
Please rename the above two functions to `normalizeLongDoubleToXDecimalInt128` and `normalizeLongDoubleToXDecimalInt` respectively.

tntnatbry commented on dbcon/joblist/tupleunion.cpp:
```diff
@@ -56,6 +56,1296 @@ inline double exp10(double x)
+    int diff = out->getScale(i) - scale;
+    ival = datatypes::applySignedScale<uint64_t>(ival, diff);
+    out->setIntField(ival, i);
+  }
```
Again, please remove the duplicated code in the above two functions. This is the same code as in the above comment for `normalizeXDoubleToXDecimal`.

tntnatbry commented on dbcon/joblist/tupleunion.cpp:
```diff
@@ -56,6 +56,1296 @@ inline double exp10(double x)
+  void normalizeLongDoubleToLongDouble(const Row& in, Row* out, uint32_t i) 
+  {
+    long double val = in.getLongDoubleField(i);
+    out->setIntField((int64_t)val, i);
```
This should be `out->setLongDoubleField(val, i);`

tntnatbry commented on dbcon/joblist/tupleunion.cpp:
```diff
@@ -56,6 +56,1296 @@ inline double exp10(double x)
+    int diff = out->getScale(i) - scale;
+    ival = datatypes::applySignedScale<uint64_t>(ival, diff);
+    out->setIntField(ival, i);
+  }
```
Please remove code duplication from the above two functions.

tntnatbry commented on dbcon/joblist/tupleunion.h:
```diff
@@ -122,8 +122,8 @@ class TupleUnion : public JobStep, public TupleDeliveryStep
-  void addToOutput(rowgroup::Row* r, rowgroup::RowGroup* rg, bool keepit, rowgroup::RGData& data);
-  void normalize(const rowgroup::Row& in, rowgroup::Row* out);
+  void addToOutput(rowgroup::Row* r, rowgroup::RowGroup* rg, bool keepit, rowgroup::RGData& data, uint32_t& tmpOutputRowCount);
+  void normalize(const rowgroup::Row& in, rowgroup::Row* out, std::vector<std::function<void(const rowgroup::Row& in, rowgroup::Row* out, uint32_t col)>>& normalizeFunctions);
```
Can you create a typedef for this vector in the global namespace in this .h file and reuse it here as well as in the .cpp?
```c++
using normalizeFunctionsT = std::vector<std::function<void(const rowgroup::Row& in, rowgroup::Row* out, uint32_t col)>>;
```

Also, I think `normalizeFunctions` should be a const ref here:
```c++
void normalize(const rowgroup::Row& in, rowgroup::Row* out, const normalizeFunctionsT& normalizeFunctions);
```

JigaoLuo committed:
[MCOL-4590] UNION Performance Improvement with the focus on the normalize functions.

This patch improves the runtime performance of UNION processing in CS, as reported JIRA issue MCOL 4590. The idea of the optimization is to infer the normalize seperate functions beforehand and perform the normalization individually later, instead of a huge switch body of all normalization. This patch also cover engineering optimization, removing the hotspots in UNION processing. After application of this patch, the normalize part takes only about 25% of the whole UNION query in our experiment avg case.

Signed-off-by: Jigao Luo <luojigao@outlook.com>

tntnatbry commented on dbcon/joblist/tupleunion.cpp:
```diff
@@ -56,6 +56,1234 @@ inline double exp10(double x)
+  void normalizeLongDoubleToLongDouble(const Row& in, Row* out, uint32_t i) 
+  {
+    long double val = in.getLongDoubleField(i);
+    out->setLongDoubleField((int64_t)val, i);
```
The typecast to `int64_t` should be removed:
```c++
out->setLongDoubleField(val, i);
```

tntnatbry commented on dbcon/joblist/tupleunion.cpp:
```diff
@@ -56,6 +56,1234 @@ inline double exp10(double x)
+  {
+    uint32_t i;
+    joblist::normalizeFunctionsT result;
+    for (i = 0; i < out->getColumnCount(); i++)
```
Can you add a linebreak before the `for` statement?

tntnatbry commented on dbcon/joblist/tupleunion.cpp:
```diff
@@ -56,6 +56,1234 @@ inline double exp10(double x)
+  }
+
+
+  void normalizeIntToStringNoScale(const Row& in, Row* out, uint32_t i) 
```
There are two linebreaks before this function, please fix.

tntnatbry commented on dbcon/joblist/tupleunion.cpp:
```diff
@@ -56,6 +56,1234 @@ inline double exp10(double x)
+  }
+
+
+  void normalizeXDoubleToLongDouble(const Row& in, Row* out, uint32_t i) 
```
There are two linebreaks before this function, please fix.

JigaoLuo commented:
# Performance Testing of This PR

## Experiment Environment

The experiments are run on the following hardware configuration:
- AWS Instance type: c5.4xlarge
- Debian 11
- 30GiB EBS SSD

## Dataset

The benchmark dataset is provided by the community: https://github.com/mariadb-corporation/mariadb-columnstore-samples/
In this work, the table *flights* is focused, which has 32 columns and over 38M tuples.

### Schema

There are details of the table *flights*:
```
MariaDB [bts]> describe flights;
--
+---------------------+-------------+------+-----+---------+-------+
\| Field               \| Type        \| Null \| Key \| Default \| Extra \|
+---------------------+-------------+------+-----+---------+-------+
\| year                \| smallint(6) \| YES  \|     \| NULL    \|       \|
\| month               \| tinyint(4)  \| YES  \|     \| NULL    \|       \|
\| day                 \| tinyint(4)  \| YES  \|     \| NULL    \|       \|
\| day_of_week         \| tinyint(4)  \| YES  \|     \| NULL    \|       \|
\| fl_date             \| date        \| YES  \|     \| NULL    \|       \|
\| carrier             \| varchar(2)  \| YES  \|     \| NULL    \|       \|
\| tail_num            \| varchar(6)  \| YES  \|     \| NULL    \|       \|
\| fl_num              \| smallint(6) \| YES  \|     \| NULL    \|       \|
\| origin              \| varchar(5)  \| YES  \|     \| NULL    \|       \|
\| dest                \| varchar(5)  \| YES  \|     \| NULL    \|       \|
\| crs_dep_time        \| varchar(4)  \| YES  \|     \| NULL    \|       \|
\| dep_time            \| varchar(4)  \| YES  \|     \| NULL    \|       \|
\| dep_delay           \| smallint(6) \| YES  \|     \| NULL    \|       \|
\| taxi_out            \| smallint(6) \| YES  \|     \| NULL    \|       \|
\| wheels_off          \| varchar(4)  \| YES  \|     \| NULL    \|       \|
\| wheels_on           \| varchar(4)  \| YES  \|     \| NULL    \|       \|
\| taxi_in             \| smallint(6) \| YES  \|     \| NULL    \|       \|
\| crs_arr_time        \| varchar(4)  \| YES  \|     \| NULL    \|       \|
\| arr_time            \| varchar(4)  \| YES  \|     \| NULL    \|       \|
\| arr_delay           \| smallint(6) \| YES  \|     \| NULL    \|       \|
\| cancelled           \| smallint(6) \| YES  \|     \| NULL    \|       \|
\| cancellation_code   \| smallint(6) \| YES  \|     \| NULL    \|       \|
\| diverted            \| smallint(6) \| YES  \|     \| NULL    \|       \|
\| crs_elapsed_time    \| smallint(6) \| YES  \|     \| NULL    \|       \|
\| actual_elapsed_time \| smallint(6) \| YES  \|     \| NULL    \|       \|
\| air_time            \| smallint(6) \| YES  \|     \| NULL    \|       \|
\| distance            \| smallint(6) \| YES  \|     \| NULL    \|       \|
\| carrier_delay       \| smallint(6) \| YES  \|     \| NULL    \|       \|
\| weather_delay       \| smallint(6) \| YES  \|     \| NULL    \|       \|
\| nas_delay           \| smallint(6) \| YES  \|     \| NULL    \|       \|
\| security_delay      \| smallint(6) \| YES  \|     \| NULL    \|       \|
\| late_aircraft_delay \| smallint(6) \| YES  \|     \| NULL    \|       \|
+---------------------+-------------+------+-----+---------+-------+
32 rows in set (0.000 sec)
 
MariaDB [bts]> select count(*) from INFORMATION_SCHEMA.COLUMNS where table_name = "flights";
+----------+
\| count(*) \|
+----------+
\|       32 \|
+----------+
1 row in set (0.003 sec)
MariaDB [bts]> select count(*) from flights;
+----------+
\| count(*) \|
+----------+
\| 38083735 \|
+----------+
1 row in set (0.166 sec)
```

### Benchmark Query Q1

The following Query Q1 is the benchmark query in our experiments and the query to be optimized.
```
MariaDB [bts]> select count(*) from (select * from flights union all select * from flights) as Q1;
--
+----------+
\| count(*) \|
+----------+
\| 76167470 \|
+----------+
1 row in set (3.011 sec)
```

### Benchmark Query Q2

The following Query Q2 has no `UNION` statement.  Ideally, Q1 should be close to 2x runtime of Q2.

```
MariaDB [bts]> select count(*) from (select * from flights) as Q2;
--
+----------+
\| count(*) \|
+----------+
\| 38083735 \|
+----------+
1 row in set (1.049 sec)
```

## Benchmark Tool

The benchmark tool & script are provided by the community: https://github.com/drrtuy/cs-docker-tools

Here is how I run the Q1: `~/cs-docker-tools/slapit$ sudo ./sysbench_wrapper.sh 1 100 slapunion.lua bts`. The `slapunion.lua bts` is loaded only with Q1.
The way to benchmark Q2 is similar.

## Q2 Performance

The average runtime of Q2 is **632.42ms**.

```
Latency histogram (values are in milliseconds)
       value  ------------- distribution ------------- count
     623.335 |**************************************** 63
     634.661 |*************                            21
     646.192 |******                                   10
     657.933 |**                                       3
     694.452 |**                                       3
 
SQL statistics:
    queries performed:
        read:                            100
        write:                           0
        other:                           0
        total:                           100
    transactions:                        100    (1.58 per sec.)
    queries:                             100    (1.58 per sec.)
    ignored errors:                      0      (0.00 per sec.)
    reconnects:                          0      (0.00 per sec.)

General statistics:
    total time:                          63.2444s
    total number of events:              100

Latency (ms):
         min:                                  623.54
         avg:                                  632.42
         max:                                  698.28
         95th percentile:                      657.93
         sum:                                63242.44

Threads fairness:
    events (avg/stddev):           100.0000/0.00
    execution time (avg/stddev):   63.2424/0.00
```

## Q1 Performance Without This PR

I benchmark with this commit https://github.com/mariadb-corporation/mariadb-columnstore-engine/commits/develop, which is the last commit and this PR is based on.

The average runtime of Q1 without the optimization of this PR is **3229.35ms**.

```
Latency histogram (values are in milliseconds)
       value  ------------- distribution ------------- count
    2828.869 |***                                      1
    2880.269 |******                                   2
    2932.602 |***********                              4
    2985.887 |**************                           5
    3040.139 |*******************************          11
    3095.377 |***********************                  8
    3151.619 |***********************                  8
    3208.883 |*******************************          11
    3267.187 |*************************************    13
    3326.551 |**************************************** 14
    3386.993 |**************************               9
    3448.533 |*****************                        6
    3511.192 |*********                                3
    3574.989 |***                                      1
    3639.945 |*********                                3
    3706.081 |***                                      1
 
SQL statistics:
    queries performed:
        read:                            100
        write:                           0
        other:                           0
        total:                           100
    transactions:                        100    (0.31 per sec.)
    queries:                             100    (0.31 per sec.)
    ignored errors:                      0      (0.00 per sec.)
    reconnects:                          0      (0.00 per sec.)

General statistics:
    total time:                          322.9365s
    total number of events:              100

Latency (ms):
         min:                                 2831.68
         avg:                                 3229.35
         max:                                 3707.17
         95th percentile:                     3511.19
         sum:                               322934.52

Threads fairness:
    events (avg/stddev):           100.0000/0.00
    execution time (avg/stddev):   322.9345/0.00
```

## Q1 Performance With This PR 

The average runtime of Q1 with the optimization of this PR is **1312.31ms**.

```
Latency histogram (values are in milliseconds)
       value  ------------- distribution ------------- count
    1280.934 |*************************                30
    1304.208 |**************************************** 48
    1327.905 |*********                                11
    1352.033 |****                                     5
    1376.599 |**                                       2
    1401.611 |*                                        1
    1427.078 |*                                        1
    1453.007 |*                                        1
    1771.289 |*                                        1
 
SQL statistics:
    queries performed:
        read:                            100
        write:                           0
        other:                           0
        total:                           100
    transactions:                        100    (0.76 per sec.)
    queries:                             100    (0.76 per sec.)
    ignored errors:                      0      (0.00 per sec.)
    reconnects:                          0      (0.00 per sec.)

General statistics:
    total time:                          131.2332s
    total number of events:              100

Latency (ms):
         min:                                 1285.18
         avg:                                 1312.31
         max:                                 1768.78
         95th percentile:                     1376.60
         sum:                               131231.23

Threads fairness:
    events (avg/stddev):           100.0000/0.00
    execution time (avg/stddev):   131.2312/0.00
```

## Summary

Q2 AVG Runtime: 0.63s
Q1 AVG Runtime: 3.23s
Optimized Q1 AVG Runtime: 1.31s

The *Runtime Slowdown Ratio* of Q1 and Q2 is ~5x which means the Q1 has more than 5 times the runtime of Q2.  Ideally, the Runtime Slowdown Ratio should be close to 2. The current `develop` branch has a very inefficient UNION processing logic with overhead.

Applying this patch, the runtime of Q1 is optimized to 1.31s, resulting in the Runtime Slowdown Ratio of 2.07. This ratio is very close to the ideal ratio. Moreover, the theoretical minimum is 2, which makes it impossible to optimize this ratio under 2.

In summary, I have optimized the UNION processing in ColumnStore. The performance improvement is satisfying and close to a theoretical limit.

Hinal-Srivastava commented:
Hello,
I am interested in contributing to this project. Can you please tell me some resources to get started and also guide me towards some first timer tasks? 
Thank you! Looking forward to hearing from you

drrtuy commented:
Hi @Hinal-Srivastava,
I am glad to hear that. Could you get in touch with me in [Zulip](https://mariadb.zulipchat.com/#narrow/stream/118759-general/topic/ColumnStore)?
